// inngest/components/execution/InngestExecution.cjs
const require_rolldown_runtime = require('../../_virtual/rolldown_runtime.cjs');
const require_consts = require('../../helpers/consts.cjs');
let debug = require("debug");
debug = require_rolldown_runtime.__toESM(debug);

//#region src/components/execution/InngestExecution.ts
var InngestExecution_exports = /* @__PURE__ */ require_rolldown_runtime.__export({
	ExecutionVersion: () => require_consts.ExecutionVersion,
	InngestExecution: () => InngestExecution,
	PREFERRED_ASYNC_EXECUTION_VERSION: () => PREFERRED_ASYNC_EXECUTION_VERSION,
	PREFERRED_CHECKPOINTING_EXECUTION_VERSION: () => PREFERRED_CHECKPOINTING_EXECUTION_VERSION
});
/**
* The preferred execution version that will be used by the SDK when handling
* brand new runs where the Executor is allowing us to choose.
*
* Changing this should not ever be a breaking change, as this will only change
* new runs, not existing ones.
*/
const PREFERRED_ASYNC_EXECUTION_VERSION = require_consts.ExecutionVersion.V1;
/**
* The preferred execution version that will be used by the SDK when handling
* checkpointed runs where the Executor is allowing us to choose.
*
* Changing this should not ever be a breaking change, as this will only change
* new runs, not existing ones.
*/
const PREFERRED_CHECKPOINTING_EXECUTION_VERSION = require_consts.ExecutionVersion.V2;
var InngestExecution = class {
	debug;
	constructor(options) {
		this.options = options;
		this.debug = (0, debug.default)(`${require_consts.debugPrefix}:${this.options.runId}`);
	}
};

//#endregion
exports.InngestExecution = InngestExecution;
Object.defineProperty(exports, 'InngestExecution_exports', {
  enumerable: true,
  get: function () {
    return InngestExecution_exports;
  }
});
exports.PREFERRED_ASYNC_EXECUTION_VERSION = PREFERRED_ASYNC_EXECUTION_VERSION;
exports.PREFERRED_CHECKPOINTING_EXECUTION_VERSION = PREFERRED_CHECKPOINTING_EXECUTION_VERSION;
